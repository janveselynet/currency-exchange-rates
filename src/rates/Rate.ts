export interface Rate {
  country: string;
  currency: string;
  currencyCode: string;
  rate: number;
}
