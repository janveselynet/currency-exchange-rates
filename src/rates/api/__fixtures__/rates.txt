25 Nov 2022 #228
Country|Currency|Amount|Code|Rate
Australia|dollar|1|AUD|15.831
Brazil|real|1|BRL|4.395
Bulgaria|lev|1|BGN|12.473
Canada|dollar|1|CAD|17.559
China|renminbi|1|CNY|3.278
Croatia|kuna|1|HRK|3.235
Hungary|forint|100|HUF|5.923
