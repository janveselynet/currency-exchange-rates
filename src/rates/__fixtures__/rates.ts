import { Rate } from '../Rate';

export const rates: Array<Rate> = [
  {
    country: 'Australia',
    currency: 'dollar',
    currencyCode: 'AUD',
    rate: 15.831,
  },
  {
    country: 'Brazil',
    currency: 'real',
    currencyCode: 'BRL',
    rate: 4.395,
  },
  {
    country: 'Bulgaria',
    currency: 'lev',
    currencyCode: 'BGN',
    rate: 12.473,
  },
  {
    country: 'Canada',
    currency: 'dollar',
    currencyCode: 'CAD',
    rate: 17.559,
  },
  {
    country: 'China',
    currency: 'renminbi',
    currencyCode: 'CNY',
    rate: 3.278,
  },
  {
    country: 'Croatia',
    currency: 'kuna',
    currencyCode: 'HRK',
    rate: 3.235,
  },
];
