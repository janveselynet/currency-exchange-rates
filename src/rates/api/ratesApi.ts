import { Rate } from '../Rate';

const CNB_RATES_URL = 'https://www.cnb.cz/en/financial-markets/foreign-exchange-market/central-bank-exchange-rate-fixing/central-bank-exchange-rate-fixing/daily.txt';
const CSV_DELIMITER = '|';

const parseCsvIntoRates = (textContent: string): Array<Rate> => {
  const rows = textContent.split("\n");
  return rows
    .slice(2, rows.length - 1)
    .map((row: string): Rate => {
      const values = row.split(CSV_DELIMITER);
      return {
        country: values[0] ?? '',
        currency: values[1] ?? '',
        amount: parseInt(values[2]) ?? 1,
        currencyCode: values[3] ?? '',
        rate: parseFloat(values[4]) ?? 1,
      };
    });
};

export const fetchRates = async (): Promise<Array<Rate>> => {
  const response = await fetch(CNB_RATES_URL);

  if (!response.ok) {
    throw new Error(`Unable to fetch currency exchange rates from CNB, HTTP status: ${response.status}`);
  }

  const textContent = await response.text();

  return parseCsvIntoRates(textContent);
};
