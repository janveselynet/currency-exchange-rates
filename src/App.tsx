import React from 'react';
import RatesTable from './rates/table/RatesTable';
import CurrencyConverter from './converter/CurrencyConverter';
import { useQuery } from '@tanstack/react-query';
import { fetchRates } from './rates/api/ratesApi';
import { Rate } from './rates/Rate';

export default function App(): JSX.Element {
  const ratesQuery = useQuery<Array<Rate>>({ queryKey: ['rates'], queryFn: fetchRates });

  if (ratesQuery.isLoading) {
    return <p>Loading rates...</p>;
  }
  if (ratesQuery.isError) {
    const errorMessage = `Error occurred when loading rates. See details: ${ratesQuery.error}`;
    return <p>{errorMessage}</p>;
  }
  if (ratesQuery.data?.length === 0) {
    return <p>No rates available, sorry.</p>;
  }

  return (
    <React.Fragment>
      <h1>Currency converter</h1>
      <CurrencyConverter availableRates={ratesQuery.data} />

      <h1>Currency exchange rates</h1>
      <RatesTable rates={ratesQuery.data} />
    </React.Fragment>
  );
}
