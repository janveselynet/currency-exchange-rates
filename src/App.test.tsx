import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import App from './App';
import * as ratesApi from './rates/api/ratesApi';
import { rates } from './rates/__fixtures__/rates';

describe('App', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const getAppElementToRender = (): JSX.Element => {
    const testQueryClient = new QueryClient({
      defaultOptions: {
        queries: { retry: false },
      },
    });

    return (
      <QueryClientProvider client={testQueryClient}>
        <App />
      </QueryClientProvider>
    );
  }

  it('renders loading state when rates are initially loading', () => {
    jest
      .spyOn(ratesApi, 'fetchRates')
      .mockImplementation(async () => new Promise(() => null));
    
    const { container } = render(getAppElementToRender());
    expect(container).toMatchSnapshot();
  });

  it('renders error state when fetching rates fails', async () => {
    jest
      .spyOn(ratesApi, 'fetchRates')
      .mockRejectedValue(new Error('Unable to fetch rates'));

    const { container } = render(getAppElementToRender());
    await waitFor(() => screen.getByText(/error/i));
    expect(container).toMatchSnapshot();
  });

  it('renders properly when no rates available', async () => {
    jest
      .spyOn(ratesApi, 'fetchRates')
      .mockResolvedValue([]);

    const { container } = render(getAppElementToRender());
    await waitFor(() => screen.getByText(/No rates available/i));
    expect(container).toMatchSnapshot();
  });

  it('renders properly when rates available', async () => {
    jest
      .spyOn(ratesApi, 'fetchRates')
      .mockResolvedValue(rates);
    
    const { container } = render(getAppElementToRender());
    await waitFor(() => screen.getByText(/Currency converter/i));
    expect(container).toMatchSnapshot();
  });
});
