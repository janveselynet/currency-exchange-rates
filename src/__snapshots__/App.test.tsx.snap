// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders error state when fetching rates fails 1`] = `
<div>
  <p>
    Error occurred when loading rates. See details: Error: Unable to fetch rates
  </p>
</div>
`;

exports[`App renders loading state when rates are initially loading 1`] = `
<div>
  <p>
    Loading rates...
  </p>
</div>
`;

exports[`App renders properly when no rates available 1`] = `
<div>
  <p>
    No rates available, sorry.
  </p>
</div>
`;

exports[`App renders properly when rates available 1`] = `
<div>
  <h1>
    Currency converter
  </h1>
  <form>
    <input
      class="sc-gswNZR cyIJhE"
      id="amount"
      placeholder="Enter amount"
      type="number"
      value=""
    />
    CZK to
    <select
      class="sc-dkrFOg hFFHwP"
      name="currencyCode"
    >
      <option
        value="AUD"
      >
        AUD
      </option>
      <option
        value="BRL"
      >
        BRL
      </option>
      <option
        value="BGN"
      >
        BGN
      </option>
      <option
        value="CAD"
      >
        CAD
      </option>
      <option
        value="CNY"
      >
        CNY
      </option>
      <option
        value="HRK"
      >
        HRK
      </option>
      <option
        value="HUF"
      >
        HUF
      </option>
    </select>
    <button
      class="sc-hLBbgP hZYYNp"
    >
      Convert
    </button>
  </form>
  <p>
    0
     CZK = 
    0
     
    AUD
  </p>
  <h1>
    Currency exchange rates
  </h1>
  <table
    class="sc-bcXHqe iPxklV"
  >
    <thead>
      <tr>
        <th>
          Country
        </th>
        <th>
          Currency
        </th>
        <th>
          Amount
        </th>
        <th>
          Code
        </th>
        <th>
          Rate
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Australia
        </td>
        <td>
          dollar
        </td>
        <td>
          1
        </td>
        <td>
          AUD
        </td>
        <td>
          15.831
        </td>
      </tr>
      <tr>
        <td>
          Brazil
        </td>
        <td>
          real
        </td>
        <td>
          1
        </td>
        <td>
          BRL
        </td>
        <td>
          4.395
        </td>
      </tr>
      <tr>
        <td>
          Bulgaria
        </td>
        <td>
          lev
        </td>
        <td>
          1
        </td>
        <td>
          BGN
        </td>
        <td>
          12.473
        </td>
      </tr>
      <tr>
        <td>
          Canada
        </td>
        <td>
          dollar
        </td>
        <td>
          1
        </td>
        <td>
          CAD
        </td>
        <td>
          17.559
        </td>
      </tr>
      <tr>
        <td>
          China
        </td>
        <td>
          renminbi
        </td>
        <td>
          1
        </td>
        <td>
          CNY
        </td>
        <td>
          3.278
        </td>
      </tr>
      <tr>
        <td>
          Croatia
        </td>
        <td>
          kuna
        </td>
        <td>
          1
        </td>
        <td>
          HRK
        </td>
        <td>
          3.235
        </td>
      </tr>
      <tr>
        <td>
          Hungary
        </td>
        <td>
          forint
        </td>
        <td>
          100
        </td>
        <td>
          HUF
        </td>
        <td>
          5.923
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
